Dynamo support for autograd.Function w/ once_differentiable

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108686

Fixes #106893

There are two main changes:
- Before this PR, the function returned by once_differentiable was
included in skipfiles (because its .co_code is
torch/autograd/function.py). This PR adds a mechanism to tell Dynamo
to inline a function, no matter if it is included in skipfiles.
- A bugfix: when we are introspecting the backward, we need to turn the
grad mode off. This is to accurately model the eager-mode semantics:
In eager-mode PyTorch, if second-order gradients were not requested, then
the grad mode is off. torch.compile does not work with higher-order
gradients and just assumes we do first-order gradients, so this is OK.

Test Plan:
- new test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Differential Revision: [D49064185](https://our.internmc.facebook.com/intern/diff/D49064185)

zou3519 committed:
Dynamo support for autograd.Function w/ once_differentiable

Fixes #106893

There are two main changes:
- Before this PR, the function returned by once_differentiable was
included in skipfiles (because its .co_code is
torch/autograd/function.py). This PR special cases Dynamo to inline the
function.
- A bugfix: when we are introspecting the backward, we need to turn the
grad mode off. This is to accurately model the eager-mode semantics:
In eager-mode PyTorch, if second-order gradients were not requested, then
the grad mode is off. torch.compile does not work with higher-order
gradients and just assumes we do first-order gradients, so this is OK.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -2207,13 +2207,30 @@ def check_inlineable(func):
                 return
+
+
+            # Function returned by torch.autograd.function.once_differentiable is known sound
+            if (
+                isinstance(func, UserFunctionVariable) and
+                func.get_filename().endswith('torch/autograd/function.py') and
+                not func.fn.__module__.startswith('torch.autograd.function')
+            ):
+                return
+            # Function wrapped by torch.autograd.function.once_differentiable is known sound
+            if (
+                isinstance(func, NestedUserFunctionVariable) and
+                func.get_filename().endswith('torch/autograd/function.py') and
+                func.fn_name.value == 'once_differentiable.<locals>.wrapper.<locals>.<genexpr>'
```
Seems super hacky. I can try the "decorator to mark a function to always be inlined" idea if we think this is too hacky.

voznesenskym replied:
Yeah as this is, it doesn't spark joy.

One thing we can do is like `wrapper.once_differentiable = True` and then just hasattr check that field? Still gross, but beats this value check on string (which is not safe anyway, if someone refactors once_differentiable lol)

What was wrong with inlining btw?

zou3519 replied:
> What was wrong with inlining btw?

Before this PR, Dynamo attempts to inline the function produced by once_differentiable. Then, it fails in `InliningInstructionTranslator.check_inlineable` and complains that the function being inlined is in skipfiles.

zou3519 committed:
Update on "Dynamo support for autograd.Function w/ once_differentiable"


Fixes #106893

There are two main changes:
- Before this PR, the function returned by once_differentiable was
included in skipfiles (because its .co_code is
torch/autograd/function.py). This PR special cases Dynamo to inline the
function.
- A bugfix: when we are introspecting the backward, we need to turn the
grad mode off. This is to accurately model the eager-mode semantics:
In eager-mode PyTorch, if second-order gradients were not requested, then
the grad mode is off. torch.compile does not work with higher-order
gradients and just assumes we do first-order gradients, so this is OK.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Dynamo support for autograd.Function w/ once_differentiable"


Fixes #106893

There are two main changes:
- Before this PR, the function returned by once_differentiable was
included in skipfiles (because its .co_code is
torch/autograd/function.py). This PR adds a mechanism to tell Dynamo
to inline a function, no matter if it is included in skipfiles.
- A bugfix: when we are introspecting the backward, we need to turn the
grad mode off. This is to accurately model the eager-mode semantics:
In eager-mode PyTorch, if second-order gradients were not requested, then
the grad mode is off. torch.compile does not work with higher-order
gradients and just assumes we do first-order gradients, so this is OK.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Dynamo support for autograd.Function w/ once_differentiable"


Fixes #106893

There are two main changes:
- Before this PR, the function returned by once_differentiable was
included in skipfiles (because its .co_code is
torch/autograd/function.py). This PR adds a mechanism to tell Dynamo
to inline a function, no matter if it is included in skipfiles.
- A bugfix: when we are introspecting the backward, we need to turn the
grad mode off. This is to accurately model the eager-mode semantics:
In eager-mode PyTorch, if second-order gradients were not requested, then
the grad mode is off. torch.compile does not work with higher-order
gradients and just assumes we do first-order gradients, so this is OK.

Test Plan:
- new test

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo_utils.py:
```diff
@@ -0,0 +1,29 @@
+DYNAMO_FORCE_INLINE: Any = weakref.WeakKeyDictionary()
+
+
+def compiler_force_inline(f):
```
nit: drop compiler from the fn names if its under _dynamo_utils?

voznesenskym commented on torch/_dynamo_utils.py:
```diff
@@ -0,0 +1,29 @@
+    in SKIPFILES for inlining. We should rework the SKIPFILES design
+    (and this API).
+    """
+    DYNAMO_FORCE_INLINE[f.__code__] = True
```
nit assert function type? even some `callable` passing objects like functools.partial won't have a __code__

voznesenskym commented on torch/_dynamo_utils.py:
```diff
@@ -0,0 +1,29 @@
+    return f
+
+
+def compiler_should_force_inline(func):
+    return func.get_code() in DYNAMO_FORCE_INLINE
```
this func is a VT, right? Let's assert that its a known VT type, and maybe move this to regular dynamo utils?

voznesenskym commented on torch/autograd/function.py:
```diff
@@ -550,7 +551,9 @@ def apply(cls, *args, **kwargs):
 
 
 def once_differentiable(fn):
+    @compiler_force_inline
     @functools.wraps(fn)
+    @compiler_force_inline
```
Comment please?

zou3519 committed:
Update on "Dynamo support for autograd.Function w/ once_differentiable"


Fixes #106893

There are two main changes:
- Before this PR, the function returned by once_differentiable was
included in skipfiles (because its .co_code is
torch/autograd/function.py). This PR adds a mechanism to tell Dynamo
to inline a function, no matter if it is included in skipfiles.
- A bugfix: when we are introspecting the backward, we need to turn the
grad mode off. This is to accurately model the eager-mode semantics:
In eager-mode PyTorch, if second-order gradients were not requested, then
the grad mode is off. torch.compile does not work with higher-order
gradients and just assumes we do first-order gradients, so this is OK.

Test Plan:
- new test

[ghstack-poisoned]

xw285cornell commented on torch/autograd/function.py:
```diff
@@ -550,7 +551,14 @@ def apply(cls, *args, **kwargs):
+    # the following decorator (once per function) to force inlines.
+    @torch._dynamo_utils.force_inline
     @functools.wraps(fn)
+    @torch._dynamo_utils.force_inline
```
this line is duplicated?

zou3519 replied:
This is intentional, we need to force_inline on `wrapper` and force_inline on the function produced by functools.wraps

zou3519 commented:
@pytorchbot merge

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49064185).

zou3519 commented:
@pytorchbot merge
